### ISSUE TITLE
echo - called party/callee can hear ownself

### ISSUE THREAD

**Issue by rabongithub** [juha-h/baresip-studio#188]
did not try yet your 35.0.0.0, but is the new module webrtc_aecm will fix the echo issue ?

**Comment by xandark** [comment]
People are telling me repeatedly that they can hear their voice echoed back to them when I call on my Pixel 7 phone. So from the previous comment, it seems as if Baresip cannot use hardware echo cancellation, is that right?

**Comment by juha-h** [comment]
Since baresip lib that handles also audio processing in written in C, it cannot use Android echo cancellation capabilities.  For echo cancellation, the lib uses webrtc audio processing lib https://gitlab.freedesktop.org/pulseaudio/webrtc-audio-processing.

Make sure that you have enabled Acoustic Echo Cancellation in audio settings.  There has not been too much echo during the calls that I have made.

**Comment by xandark** [comment]
Okay interesting that being written in C precludes using it, never would've thought that. 

Yes, that setting has been enabled and I've triple-checked it because I thought there had to be something I was missing. Well, I will have to find another app, I wish I didn't, I really like everything about Baresip. But thank you for your replies.

**Comment by juha-h** [comment]
It is not really C, but audio processing done by baresip lib API, not Android audio API. Yes, you need to use some other app, e.g. linphone.

**Comment by juha-h** [comment]
@rabongithub @xandark There new hope for better echo control, automatic gain control, and noise suppression via the new AAudio module that is currently being worked on.  It gives the app access to audio sessionId, which then allows using those audio effects on baresip app.

I just created a test app at https://box.tutpro.com/tmp/baresip.apk that uses the new audio module. Please give it a try (backup before installing and restore after).

**Comment by xandark** [comment]
Cool! Will do!

**Comment by xandark** [comment]
Okay, I've installed this on my Pixel 7 but it's very difficult to use, the app extends its display up underneath the top system area with the time and the system icons. And it also extends below the bottom where I have my back, home, and menu icons:

![my phone](https://github.com/user-attachments/assets/1dcf5f52-e10c-49b6-92d9-ed88e3c43b28)

So it is very hard to trigger the three dot menu in the upper right hand corner. Is there some setting I need to do on my end?

**Comment by juha-h** [comment]
xandark writes:

> Okay, I've installed this on my Pixel 7 but it's very difficult to
> use, the app extends its display up underneath the top system area
> with the time and the system icons. And it also extends below the
> bottom where I have my back, home, and menu icons:

I cannot reproduce that on my device or Google Pixel emulator.  Which
device/OS you have?

-- Juha

**Comment by juha-h** [comment]
Here is screenshot of the app running on Pixel emulator and all looks fine:

![shot](https://github.com/user-attachments/assets/211fc60a-7e58-4890-b10d-046b456123f4)

**Comment by juha-h** [comment]
@xandark Do you have official Android OS on your device or something
else?

If yes, how come the official Pixel 7 API 34 emulator does not show your
issue?

**Comment by xandark** [comment]
Yes, completely official OS on my device, fully updated to latest version. I have not seen other apps do this where they take up the entirety of the screen. Not sure why the emulator doesn't show the issue. In any case, I'll try to test the new feature.

**Comment by juha-h** [comment]
Is it only this apk that you downloaded where is screen issue appears or does it appear also on the latest published version 61.1.0 available from F-Droid?

**Comment by xandark** [comment]
Only this downloaded version. I've never experienced that with any of the ones from F-Droid. 
I tried to use this version today on my Pixel 7, but because of the UI issues I'm experiencing, I just could not make any headway. I'll have to try on an older phone.

**Comment by juha-h** [comment]
xandark writes:

> Only this downloaded version. I've never experienced that with any of
> the ones from F-Droid.

It has no changes related to activity layouts as compared to published
one.  Wait until a new version (62.0.0) gets published on F-Droid and
try with that.

**Comment by xandark** [comment]
> @rabongithub @xandark There new hope for better echo control, automatic gain control, and noise suppression via the new AAudio module that is currently being worked on. It gives the app access to audio sessionId, which then allows using those audio effects on baresip app.
> 
> I just created a test app at https://box.tutpro.com/tmp/baresip.apk that uses the new audio module. Please give it a try (backup before installing and restore after).

Okay! I am now using and testing Baresip 63.1.0 from F-droid on my Pixel 7/Android 15. Using speakerphone, the other party is immediately hearing their own voice, so the noise suppression via the AAudio module doesn't seem to be working right now. I'm exclusively using the g729 codec. 

In the Audio settings page in Baresip, do I need to change the Audio Delay setting? Right now, the default seems to be 500. Which seems kind of high... Mic gain is 1.0.

How can I give you better feedback to diagnose this?

**Comment by juha-h** [comment]
Audio Delay is not related to echo.  If you see Mic Gain in Audio settings, it means that your device does not support Automatic Gain Control:
```
if (!BaresipService.agc) {
            micGain = binding.MicGain
            oldMicGain = Config.variable("augain")
            micGain.setText(oldMicGain)
        } else {
            binding.MicGainLayout.visibility = View.GONE
        }
```
Perhaps it does not support Acoustic Echo Canceler either.  You can see with logcat what the situation is by turning on Debug setting and and then restarting baresip.

**Comment by xandark** [comment]
Okay, thank you for making it easy for me to do this. I thought I had to remember how to get to it via adb. This was much easier. The result I got was:

01-21 20:37:27.771 17951 17951 D Baresip : AEC/AGC/NS available = true/false/true

This confirms my experience with this phone in that every other communication app I've used, six off the top of my head, worked flawlessly in speakerphone mode with zero feedback. This is a flagship phone.

**Comment by xandark** [comment]
Yes, I'm open to that, but the last time we tried this in November, the apk rendered over my phone's system UI and I couldn't do anything, I couldn't test it. This is what I said before:

> Okay, I've installed this on my Pixel 7 but it's very difficult to use, the app extends its display up underneath the top system area with the time and the system icons. And it also extends below the bottom where I have my back, home, and menu icons:
> 
> ![my phone](https://github.com/user-attachments/assets/1dcf5f52-e10c-49b6-92d9-ed88e3c43b28)
> 
> So it is very hard to trigger the three dot menu in the upper right hand corner. Is there some setting I need to do on my end?

**Comment by juha-h** [comment]
The padding issue should be fixed in already in version 63.0.0 (Dec 15).  I created new version 63.2.0 where AEC is applied only to recorder session.  I don't know if it helps.  The release is available here: https://github.com/juha-h/baresip-studio/releases/tag/63.2.0.  Backup before install.

**Comment by xandark** [comment]
Okay, I can try it, but the files at the link you mentioned are source files, I don't have a dev setup to build those into an apk.

**Comment by juha-h** [comment]
xandark writes:

> Okay, I can try it, but the files at the link you mentioned are source
> files, I don't have a dev setup to build those into an apk.

Something went wrong and the new release became draft.  Here is working
link:

https://github.com/juha-h/baresip-studio/releases/download/63.2.0/app-release.apk

**Comment by juha-h** [comment]
@rabongithub @xandark Latest baresip version 63.2.5 gives an alert warning if hardware AEC is not supported when user turns off Acoustic Echo Cancellation in Audio Settings.  Please try with your devices.

My cheap Samsung phone claims that AEC is available when baresip checks it with `AcousticEchoCanceler.isAvailable()` function call, but then when baresip tries to create AEC on the recording stream opened in aaudio, it produces error messages:
```
AudioEffect             com.tutpro.baresip                   E  set(): AudioFlinger could not create effect 7b491460-8d4d-11e0-bd61-0002a5d5c51b / ec7178ec-e5e1-4432-a3f4-4657e6795210, status: -22
AudioEffects-JNI        com.tutpro.baresip                   E  AudioEffect initCheck failed -3
AudioEffect-JAVA        com.tutpro.baresip                   E  Error code -3 when initializing AudioEffect.
AcousticEchoCanceler    com.tutpro.baresip                   W  not enough memory
```
According to
https://developer.android.com/ndk/reference/group/audio#aaudiostreambuilder_setsessionid
it should be possible to use audio effects on aaudio produced sessionId.

**Comment by xandark** [comment]
Okay, I'll try this when I can, but maybe more people would be able to participate in this if this issue wasn't marked as Closed. It's actually a miracle that I found this issue in the first place.

**Comment by juha-h** [comment]
This issue is very old.  If device does not support hardware AEC on audio stream opened by aaudio, there is nothing I can do to improve AEC.  baresip is then using WebRTC software based AEC, which may not work well at all.

If your device does not support not support hardware AEC on an audio stream opened by aaudio and instead produces the above error messages (adb logcat | grep com.tutpro.baresip), you need to open an issue with your device manufacturer and quote the above cited Android reference.

**Comment by xandark** [comment]
Just to follow up on this thread where we couldn't figure out whey AEC wasn't working on my Pixel 7. I just installed 66.0.0 and was very pleased to discover that AEC seems to be working now on my Pixel 7, at least when I dialed into the Echo test. I installed the package right from the github downloads page, but Android wouldn't let me enable the Allow display over other apps and the Default phone app permissions on a non-store package, so I had to install from F-Droid, which is at 65.1.0, two versions behind. I'm already noticing that there are some minor annoyances with it that were fixed in 66.0.0. Could you push the latest to F-Droid? That would be great!

**Comment by juha-h** [comment]
Good to hear that AEC works in your device with baresip 66.0.0.  Same with my device.  Its build is in F-Droid queue.  Will take a few days to build and then publish.
